### PR TITLE
[mlir][spirv] Add spirv-to-llvm conversion for group operations

### PR DESCRIPTION
**Pull request by sommerlukas**
Lowering for some of the uniform and non-uniform group operations defined in section 3.52.21 of the SPIR-V specification from SPIR-V dialect to LLVM dialect.

Similar to #111864, lower the operations to builtin functions understood by SPIR-V tools.

### CODE REVIEW

File: mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp [hunk 1]
```diff
@@ -1089,6 +1096,186 @@ class ControlBarrierPattern
...
 };
 
+namespace {
+
+StringRef getTypeMangling(Type type, bool isSigned) {
+  return llvm::TypeSwitch<Type, StringRef>(type)
+      .Case<Float16Type>([](auto) { return "Dh"; })
+      .template Case<Float32Type>([](auto) { return "f"; })
+      .template Case<Float64Type>([](auto) { return "d"; })
+      .template Case<IntegerType>([isSigned](IntegerType intTy) {
```
Comment (kuhar) [inline mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp:1106]: I don't think we need these `.template` here; the type switch is not a dependent name

File: mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp [hunk 2]
```diff
@@ -1089,6 +1096,186 @@ class ControlBarrierPattern
...
+        assert(false && "No mangling defined");
+        return "";
+      });
+}
+
+template <typename ReduceOp>
+constexpr StringLiteral getGroupFuncName() {
+  assert(false && "No builtin defined");
+  return "";
+}
```
Comment (victor-eds) [inline mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp:1134]: Not adding this in the first place is a compilation error. Better IMO.

**Review by FMarno — COMMENTED** [review]
just nits really

File: mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp [hunk 3]
```diff
@@ -1722,7 +1909,50 @@ void mlir::populateSPIRVToLLVMConversionPatterns(
...
       // Barrier ops
-      ControlBarrierPattern>(patterns.getContext(), typeConverter);
+      ControlBarrierPattern,
+
+      // Group reduction operations
+      GroupReducePattern<spirv::GroupIAddOp>,
+      GroupReducePattern<spirv::GroupFAddOp>,
+      GroupReducePattern<spirv::GroupFMinOp>,
+      GroupReducePattern<spirv::GroupUMinOp>,
+      GroupReducePattern<spirv::GroupSMinOp, /*Signed*/ true>,
```
Comment (victor-eds) [inline mlir/lib/Conversion/SPIRVToLLVM/SPIRVToLLVM.cpp:1919]:
```suggestion
      GroupReducePattern<spirv::GroupSMinOp, /*Signed=*/true>,
```

`clang-format` likes this better (same for below)

**Review by victor-eds — COMMENTED** [review]
LGTM but nits

**Review by victor-eds — APPROVED** [review]

**Comment by kuhar** [comment]
(no objections from me)

**Review by FMarno — APPROVED** [review]

**Pull request merged**